compiler/src/dotty/tools/dotc/typer/Applications.scala

```diff
@@ -631,12 +631,31 @@ trait Applications extends Compatibility { self: Typer with Dynamic =>
       def simpleApply(fun1: Tree, proto: FunProto)(implicit ctx: Context): Tree =
         methPart(fun1).tpe match {
           case funRef: TermRef =>
-            val app =
-              if (proto.allArgTypesAreCurrent())
-                new ApplyToTyped(tree, fun1, funRef, proto.typedArgs, pt)
-              else
-                new ApplyToUntyped(tree, fun1, funRef, proto, pt)(argCtx(tree))
-            convertNewGenericArray(ConstFold(app.result))
+            if (ctx.macrosEnabled && macros.isQuasiquote(funRef.termSymbol, tree))
+              typed(macros.expandQuasiquote(tree, isTerm = true), pt)
+            else if (ctx.macrosEnabled && macros.isMetaApply(funRef.termSymbol) && ctx.owner.isInlineMethod) {
+              // type meta block
+              val typer = new macros.Typer
+              fun1.appliedTo(typer.typed(tree.args.head, macros.metaTreeType)(ctx.fresh.setTyper(typer)))
+            }
+            else {
+              val app =
+                if (proto.allArgTypesAreCurrent())
+                  new ApplyToTyped(tree, fun1, funRef, proto.typedArgs, pt)
+                else
+                  new ApplyToUntyped(tree, fun1, funRef, proto, pt)(argCtx(tree))
+              val treeTpd = convertNewGenericArray(ConstFold(app.result))
+
+              // expand def macros after type checking
+              // Note: macros that expand to functions cannot be immediately applied.
+              // This is a practical constraint, as currying is also disallowed for macros
+              if (fun1.symbol.is(Flags.Macro) && !pt.isInstanceOf[ApplyingProto]) {
+                if (ctx.macrosEnabled)
+                  typed(macros.expandDef(treeTpd), pt)
```

Comment:
Am I right in thinking that you're using `expandDef` to expand def macros? Maybe we could think of a different name, because `expandDef` can be thought to mean `expand definition`?